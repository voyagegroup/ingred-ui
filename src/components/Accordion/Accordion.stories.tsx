import * as React from "react";
import { Story } from "@storybook/react/types-6-0";
import Accordion, { AccordionProps } from "./Accordion";

export default {
  title: "Components/navigation/Accordion",
  component: Accordion,
  args: {
    title: "title",
    children: "Lorem ipsum dolor sit amet",
    disabled: false,
    expanded: false,
  },
};

export const Example: Story<AccordionProps> = (args) => (
  <>
    <Accordion
      title={args.title}
      disabled={args.disabled}
      expanded={args.expanded}
    >
      <div style={{ padding: "4px 8px" }}>{args.children}</div>
    </Accordion>
  </>
);

export const Multiple: Story<AccordionProps> = () => (
  <>
    <Accordion title="title1">
      <div style={{ padding: "4px 8px" }}>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc velit
        magna, dictum vel elementum ac, fermentum quis lorem. Aliquam sapien
        augue, efficitur eget tincidunt non, accumsan et est. Aliquam lobortis
        elit lorem, id ullamcorper libero elementum non. Sed rutrum porta enim,
        et aliquet velit hendrerit ac. Morbi tincidunt eleifend elit.
      </div>
    </Accordion>
    <Accordion title="title2">
      <div style={{ padding: "4px 8px" }}>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc velit
        magna, dictum vel elementum ac, fermentum quis lorem. Aliquam sapien
        augue, efficitur eget tincidunt non, accumsan et est. Aliquam lobortis
        elit lorem, id ullamcorper libero elementum non. Sed rutrum porta enim,
        et aliquet velit hendrerit ac. Morbi tincidunt eleifend elit.
      </div>
    </Accordion>
    <Accordion disabled title="disabled">
      <div style={{ padding: "4px 8px" }}>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc velit
        magna, dictum vel elementum ac, fermentum quis lorem. Aliquam sapien
        augue, efficitur eget tincidunt non, accumsan et est. Aliquam lobortis
        elit lorem, id ullamcorper libero elementum non. Sed rutrum porta enim,
        et aliquet velit hendrerit ac. Morbi tincidunt eleifend elit.
      </div>
    </Accordion>
  </>
);

export const Controlled: Story<AccordionProps> = () => {
  const [expanded, setExpanded] = React.useState<string | false>(false);
  const handleChange =
    (panel: string) => (event: React.SyntheticEvent, newExpanded: boolean) => {
      setExpanded(newExpanded ? panel : false);
    };

  return (
    <>
      <Accordion
        title="title1"
        expanded={expanded === "title1"}
        onChange={handleChange("title1")}
      >
        <div style={{ padding: "4px 8px" }}>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc velit
          magna, dictum vel elementum ac, fermentum quis lorem. Aliquam sapien
          augue, efficitur eget tincidunt non, accumsan et est. Aliquam lobortis
          elit lorem, id ullamcorper libero elementum non. Sed rutrum porta
          enim, et aliquet velit hendrerit ac. Morbi tincidunt eleifend elit.
        </div>
      </Accordion>
      <Accordion
        title="title2"
        expanded={expanded === "title2"}
        onChange={handleChange("title2")}
      >
        <div style={{ padding: "4px 8px" }}>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc velit
          magna, dictum vel elementum ac, fermentum quis lorem. Aliquam sapien
          augue, efficitur eget tincidunt non, accumsan et est. Aliquam lobortis
          elit lorem, id ullamcorper libero elementum non. Sed rutrum porta
          enim, et aliquet velit hendrerit ac. Morbi tincidunt eleifend elit.
        </div>
      </Accordion>
      <Accordion
        title="title3"
        expanded={expanded === "title3"}
        onChange={handleChange("title3")}
      >
        <div style={{ padding: "4px 8px" }}>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc velit
          magna, dictum vel elementum ac, fermentum quis lorem. Aliquam sapien
          augue, efficitur eget tincidunt non, accumsan et est. Aliquam lobortis
          elit lorem, id ullamcorper libero elementum non. Sed rutrum porta
          enim, et aliquet velit hendrerit ac. Morbi tincidunt eleifend elit.
        </div>
      </Accordion>
    </>
  );
};
