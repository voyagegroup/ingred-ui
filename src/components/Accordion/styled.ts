import Flex from "../Flex";
import styled from "styled-components";

export const AccordionTitle = styled(Flex)<{
  expanded: boolean;
  disabled?: boolean;
}>`
  background-color: ${({ expanded, disabled, theme }) =>
    // eslint-disable-next-line no-nested-ternary
    disabled
      ? theme.palette.gray.light
      : expanded
      ? theme.palette.primary.highlight
      : theme.palette.gray.highlight};
  color: ${({ disabled, theme }) =>
    disabled ? theme.palette.text.disabled : "auto"};
  cursor: ${({ disabled }) => (disabled ? "not-allowed" : "pointer")};
  pointer-events: ${({ disabled }) => (disabled ? "none" : "auto")};
  justify-content: space-between;
  border-top: ${({ expanded, theme }) =>
    !expanded ? `1px solid ${theme.palette.divider}` : "none"};
  transition:
    background-color 0.3s,
    border-top 0.3s;
`;

export const AccordionTitleChildren = styled.div`
  padding: ${({ theme }) => theme.spacing * 1.25}px;
`;

export const IconContainer = styled.div`
  display: flex;
  box-sizing: border-box;
`;

export const IconButton = styled.div<{ expanded: boolean }>`
  padding: ${({ theme }) => `${theme.spacing * 1.25}px ${theme.spacing * 2}px`};
  transition: transform 150ms;
  transform-origin: center center;
  transform: ${({ expanded }) => (expanded ? "rotate(180deg)" : "rotate(0)")};
`;

export const DropdownIndicator = styled.div`
  position: relative;
  display: flex;
  align-items: center;
  justify-content: flex-end;
  flex: 0 0 auto;
  width: ${({ theme }) => theme.spacing * 5}px;
`;

export const AccordionContent = styled.div<{
  expanded: boolean;
  height: number;
}>`
  height: ${({ height }) => height}px;
  overflow: hidden;
  transition: 0.3s all;
  background-color: ${({ theme }) => theme.palette.white};
`;
