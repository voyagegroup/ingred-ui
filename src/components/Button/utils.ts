import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
