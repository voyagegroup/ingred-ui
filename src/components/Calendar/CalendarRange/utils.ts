import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
