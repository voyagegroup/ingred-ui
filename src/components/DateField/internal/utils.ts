export const getInputWidth = (input: string) =>
  [...input.split("")].reduce((acc, char) => {
    const code = char.charCodeAt(0);
    return acc + (code >= 0x3000 && code <= 0xff60 ? 2 : 1);
  }, 0);
