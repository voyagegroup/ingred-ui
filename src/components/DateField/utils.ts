import { useMemo } from "react";

type Sections = {
  start: number;
  end: number;
  value: string;
  editable: boolean;
};

/**
 * 何らかのフォーマットで入ってくる日付を開始位置と終了位置と値を持つセクションに分割する
 * useDateField で format された日付操作を汎用的に行うために必要なプロパティを返す
 * 例) 2023-01-02 -> [
 *   { start: 0, end: 3, value: "2023", editable: true },
 *   { start: 4, end: 4, value: "-", editable: false },
 *   { start: 5, end: 6, value: "01", editable: true },
 *   { start: 7, end: 7, value: "-", editable: false },
 *   { start: 8, end: 9, value: "02", editable: true }
 * ]
 *
 * @param formattedDate 何らかのフォーマットで入ってくる日付
 * @returns 開始位置と終了位置と値を持つセクション
 */
export const getSections = (formattedDate?: string | null) => {
  if (!formattedDate) {
    return [];
  }

  const sections: Sections[] = [];
  let start = 0;
  let isPrevCharDigit = !isNaN(Number(formattedDate[0]));

  for (let index = 1; index <= formattedDate.length; index++) {
    const currentChar = formattedDate[index];
    const isCurrentCharDigit =
      !isNaN(Number(currentChar)) && currentChar !== " ";

    if (
      isCurrentCharDigit !== isPrevCharDigit ||
      currentChar === " " ||
      index === formattedDate.length
    ) {
      sections.push({
        start,
        end: index - 1,
        value: formattedDate.slice(start, index),
        editable: isPrevCharDigit,
      });
      start = index;
      isPrevCharDigit = isCurrentCharDigit;
    }
  }

  return sections;
};

/**
 * 開始位置と終了位置と値を持つセクションをフォーマットされた日付に変換する
 */
export const formatString = (sectionsWithCharactor: Sections[]) =>
  sectionsWithCharactor.map((section) => section.value).join("");

type ReactRef<T> =
  | React.RefCallback<T>
  | React.MutableRefObject<T>
  | React.ForwardedRef<T>
  | string
  | null
  | undefined;

// from: https://github.com/voyagegroup/ingred-ui/blob/master/src/hooks/useMergeRefs.ts
export function useMergeRefs<T>(...refs: ReactRef<T>[]): React.Ref<T> {
  return useMemo(() => {
    if (refs.every((ref) => ref === null)) {
      return null;
    }
    return (refValue: T) => {
      for (const ref of refs) {
        if (typeof ref === "function") {
          ref(refValue);
        } else if (ref && typeof ref !== "string") {
          ref.current = refValue;
        }
      }
    };
  }, [refs]);
}
