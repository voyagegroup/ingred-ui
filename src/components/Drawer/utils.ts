/**
 * 相対値をpx値に変換するヘルパー関数
 * @param value - 変換する値（数値またはvw/vh/%/px等の単位付き文字列）
 * @param direction - Drawerの方向（変換時の基準値決定に使用）
 * @returns px値（数値）
 */
export const convertToPixels = (value: string | number, direction: "left" | "right" | "bottom"): number => {
  if (typeof value === "number") {
    return value; // 数値はpx扱い
  }
  
  // 文字列の場合は単位チェック
  const numericValue = parseFloat(value);
  
  if (value.includes("vw")) {
    return (numericValue / 100) * window.innerWidth;
  }
  if (value.includes("vh")) {
    return (numericValue / 100) * window.innerHeight;
  }
  if (value.includes("%")) {
    // %の場合は親要素基準だが、ここではviewport基準として扱う
    const isHorizontal = direction === "left" || direction === "right";
    return (numericValue / 100) * (isHorizontal ? window.innerWidth : window.innerHeight);
  }
  if (value.includes("px")) {
    return numericValue;
  }
  
  // 単位なしの文字列は警告を出す
  if (!isNaN(numericValue)) {
    console.warn(`Drawer: 文字列サイズには単位を指定してください。"${value}" → "${value}px" を推奨`);
    return numericValue; // 一応数値として扱う
  }
  
  // 無効な値の場合はエラー警告とフォールバック
  console.error(`Drawer: 無効なサイズ値です: "${value}". デフォルト値を使用します。`);
  return 400; // フォールバック
}; 