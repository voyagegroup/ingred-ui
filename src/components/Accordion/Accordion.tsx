import React, {
  ReactNode,
  forwardRef,
  useEffect,
  useMemo,
  useRef,
  useState,
} from "react";
import { useTheme } from "../../themes";
import Icon from "../Icon";
import * as Styled from "./styled";

export type AccordionProps = {
  title: ReactNode;
  expanded?: boolean;
  disabled?: boolean;
  onChange?: (event: React.SyntheticEvent, expanded: boolean) => void;
  children: ReactNode | ReactNode[];
};

const Accordion = forwardRef<HTMLDivElement, AccordionProps>(function Accordion(
  { title, expanded = false, disabled = false, onChange, children, ...rest },
  ref,
) {
  const theme = useTheme();
  const [expandedState, setExpandedState] = useState(expanded);
  const accordionContentContainerRef = useRef<HTMLDivElement>(null);

  const height = useMemo(() => {
    if (expandedState && accordionContentContainerRef.current) {
      return accordionContentContainerRef.current.scrollHeight;
    }
    return 0;
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [expandedState, accordionContentContainerRef.current]);

  const handleChange = (event: React.SyntheticEvent, expanded: boolean) => {
    setExpandedState(expanded);
    onChange && onChange(event, expanded);
  };

  const handleClick = (event: React.MouseEvent<HTMLDivElement>) => {
    handleChange(event, !expandedState);
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLDivElement>) => {
    if (event.key === "Enter") {
      handleChange(event, !expandedState);
    }
  };

  useEffect(() => {
    setExpandedState(expanded);
  }, [expanded]);

  return (
    <Styled.Container ref={ref} {...rest}>
      <Styled.AccordionTitle
        display="flex"
        tabIndex={disabled ? -1 : 0}
        role="button"
        aria-expanded={expandedState}
        aria-controls={`accordion-content-${title}`}
        expanded={expandedState}
        disabled={disabled}
        onClick={handleClick}
        onKeyDown={handleKeyDown}
      >
        <Styled.AccordionTitleChildren>{title}</Styled.AccordionTitleChildren>
        <Styled.DropdownIndicator>
          <Styled.IconButton expanded={expandedState}>
            <Icon
              name="arrow_bottom"
              size="md"
              color={disabled ? theme.palette.text.disabled : "black"}
            />
          </Styled.IconButton>
        </Styled.DropdownIndicator>
      </Styled.AccordionTitle>
      <Styled.AccordionContent
        ref={accordionContentContainerRef}
        expanded={expandedState}
        height={height}
      >
        {children}
      </Styled.AccordionContent>
    </Styled.Container>
  );
});

export default Accordion;
