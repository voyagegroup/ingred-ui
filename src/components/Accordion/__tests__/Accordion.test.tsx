import * as React from "react";
import "@testing-library/jest-dom";
import { cleanup } from "@testing-library/react";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";
import Accordion from "../";

describe("Accordion component testing", () => {
  afterEach(cleanup);

  test("Accordion", () => {
    const { asFragment } = renderWithThemeProvider(
      <Accordion title="Accordion Title">Accordion Content</Accordion>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("Disable Accordion", () => {
    const { asFragment } = renderWithThemeProvider(
      <Accordion disabled title="Accordion Title">
        Accordion Content
      </Accordion>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
