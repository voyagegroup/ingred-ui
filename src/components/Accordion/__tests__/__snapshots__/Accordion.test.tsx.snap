// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion component testing Accordion 1`] = `
<DocumentFragment>
  <div>
    <div
      aria-controls="accordion-content-Accordion Title"
      aria-expanded="false"
      class="sc-gsnTZi sc-dkzDqf jUiVsm cyzAVp"
      display="flex"
      role="button"
      tabindex="0"
    >
      <div
        class="sc-hKMtZM hWStLy"
      >
        Accordion Title
      </div>
      <div
        class="sc-gKXOVf iwyGgb"
      >
        <div
          class="sc-jSMfEi jmkEMm"
        >
          <span
            class="sc-bczRLJ gJuzRf"
            size="18"
          >
            <svg
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0,0H24V24H0Z"
                fill="none"
              />
              <path
                d="M13.414,16.886,7.757,11.228,9.644,9.343l3.771,3.772,3.771-3.772,1.887,1.885Z"
                fill="black"
                transform="translate(-1.414 -1.114)"
              />
            </svg>
          </span>
        </div>
      </div>
    </div>
    <div
      class="sc-iBkjds cEjDqL"
      height="0"
    >
      Accordion Content
    </div>
  </div>
</DocumentFragment>
`;

exports[`Accordion component testing Disable Accordion 1`] = `
<DocumentFragment>
  <div>
    <div
      aria-controls="accordion-content-Accordion Title"
      aria-expanded="false"
      class="sc-gsnTZi sc-dkzDqf dfqrSd SuCNp"
      disabled=""
      display="flex"
      role="button"
      tabindex="-1"
    >
      <div
        class="sc-hKMtZM hWStLy"
      >
        Accordion Title
      </div>
      <div
        class="sc-gKXOVf iwyGgb"
      >
        <div
          class="sc-jSMfEi jmkEMm"
        >
          <span
            class="sc-bczRLJ gJuzRf"
            size="18"
          >
            <svg
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0,0H24V24H0Z"
                fill="none"
              />
              <path
                d="M13.414,16.886,7.757,11.228,9.644,9.343l3.771,3.772,3.771-3.772,1.887,1.885Z"
                fill="#B3BAC1"
                transform="translate(-1.414 -1.114)"
              />
            </svg>
          </span>
        </div>
      </div>
    </div>
    <div
      class="sc-iBkjds cEjDqL"
      height="0"
    >
      Accordion Content
    </div>
  </div>
</DocumentFragment>
`;
